core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala

```diff
@@ -98,6 +98,11 @@ private[ui] class ApplicationPage(parent: MasterWebUI) extends WebUIPage("app")
                     <a href={UIUtils.makeHref(parent.master.reverseProxy,
                       app.id, app.desc.appUiUrl)}>Application Detail UI</a>
                 </strong></li>
+              } else if (parent.master.historyServerUrl.nonEmpty) {
+                <li><strong>
+                    <a href={s"${parent.master.historyServerUrl.get}/history/${app.id}"}>
```

Comment:
The history server is configured application by application, not spark master. So, IMO, we shall get the history server from the app itself instead.

Reply:
Thank you for review, @yaooqinn .

That's the reason why this is described as the best effort approach which depends on the user's setting.

However, technically, your suggestion is not true, @yaooqinn , because SHS is orthogonal from apps too. For example, app knows only `spark.eventLog.dir`, not SHS info. In addition, you can launch multiple SHS based on the same event location.
> So, IMO, we shall get the history server from the app itself instead.

Reply:
> app knows only spark.eventLog.dir, not SHS info

Not really. There is already an existing behavior for Spark on Yarn. We can set the history server via `spark.yarn.historyServer.address`

Reply:
Well, it's YARN-specific thing. I don't think that could be general in this era.

Reply:
This PR is a general solution and allows the user to provide their YARN URL to this new `spark.ui.historyServerUrl` too. I don't think this PR rejects the existing conf.

Reply:
If the user sets a different value for `spark.eventLog.dir` from the default value 
 which `spark.ui.historyServerUrl` points to, he/she might get confused while not able to find their apps

Reply:
I understand what's your point.
- However, in the production environments, SRE or submission proxy enforces it by rules in general.
- IIUC, `spark.yarn.historyServer.address` has the same issue because it depends on runtime YARN environment substitution. Did I misunderstand?

https://github.com/apache/spark/blob/b9d379a6b84b67b29ccb578938764b888d64f293/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala#L946-L955

Reply:
It seems that I found what was the conflicts here. For the following, actually, we have a team-wide small Spark cluster, not a giant one you are supposed here.
> Setting spark.eventLog.dir/spark.ui.historyServerUrl in pairs is practical, for example,
> A user wants to keep its runtime data private

Reply:
- To be clear, I agree with you that your point is valid and reasonable in a single large shared cluster.
- However, for small team-owned small clusters which belong to a single K8s cluster, the use cases and requirements are quite different what you assume.

Reply:
Only variables(spark.yarn.historyServer.address) that match the ${hadoopconf- .. } pattern are substituted. Otherwise, it's an identity operation.

Besides, `yarnConf` is client-side prioritized.

Reply:
How about we get this address from the user's app first, and if not, then from the standalone master's config?

It looks like meets the requirements for both large and small deployments.

Reply:
For this one, I can file a JIRA for that.
> How about we get this address from the user's app first, and if not, then from the standalone master's config?